### PR TITLE
Fixes #30 #31: Fixed Cluster creation/deletion. Added GCS Bucket support

### LINKED ISSUE #30: Fix Cluster creation/deletion
The cluster needs to be created beforehand as the `createCluster` and `destroyCluster` subcommands are currently not working.

### PR DESCRIPTION
**Pull request by pozgo**
### Changes:

Terraform config:
- Tidy up variables 
- Removed not needed variables
- Bumped instance type to be enough for spinning up applications.
- Disable Elasticsearch `node-pool`

Create/Destroy Cluster:
- Fixed `createCluster` and `destroyCluster` functions
- Exploded Creation to separate plan and apply for `clusterCreate`
- Added `—noop` flag to be able to run a terraform plan like mode
- Disabled `InstallHelm` (Needs rewrite)
- Added handler to check if terraform was initiated before

GCP:
- Added GCS Bucket storage creation option
- Added handler for getting the SA credentials `json` file path

Chore:
- Updated `repo` path to match it's new home (moved to `iac-io`)

---

### Tests:

Create Cluster
```
This executable full path directory is: /Users/przemek/git-repo/polinux/3rdparty/myiac
Validating flags for createCluster
2021/01/05 15:03:05 Validating flag project
2021/01/05 15:03:05 Read flag project as ozzy-playground
createCluster running with flags
2021/01/05 15:03:06 Bucket ozzy-playground-tfstate-dev exists.
Working dir is: /Users/przemek/git-repo/polinux/3rdparty/myiac/internal/terraform/cluster
Executing [ terraform apply -var-file=/Users/przemek/git-repo/polinux/3rdparty/myiac/internal/terraform/cluster/cluster.tfvars -auto-approve ]
data.terraform_remote_state.state: Refreshing state...
google_compute_network.gke_network: Creating...
google_compute_network.gke_network: Still creating... [10s elapsed]
google_compute_network.gke_network: Creation complete after 12s [id=projects/ozzy-playground/global/networks/ozzy-playground-dev-gke-network]
google_compute_firewall.gke_nodeport_service_rule: Creating...
google_compute_subnetwork.gke_subnet: Creating...
google_compute_firewall.gke_nodeport_service_rule: Still creating... [10s elapsed]
google_compute_subnetwork.gke_subnet: Still creating... [10s elapsed]
google_compute_firewall.gke_nodeport_service_rule: Creation complete after 12s [id=projects/ozzy-playground/global/firewalls/gke-nodeport-firewall-rule]
google_compute_subnetwork.gke_subnet: Creation complete after 12s [id=projects/ozzy-playground/regions/europe-west1/subnetworks/ozzy-playground-dev-gke-subnet]
google_container_cluster.cluster: Creating...
google_container_cluster.cluster: Still creating... [10s elapsed]
google_container_cluster.cluster: Still creating... [20s elapsed]

...
Apply complete! Resources: 5 added, 0 changed, 0 destroyed.


Kubernetes cluster created through Terraform from /Users/przemek/git-repo/polinux/3rdparty/myiac/internal/terraform/cluster
```
Screenshot 2021-01-05 at 15 27 16
Screenshot 2021-01-05 at 15 27 28

---
Delete Cluster

```
Waiting 5 seconds before destroying cluster...
Working dir is: /Users/przemek/git-repo/polinux/3rdparty/myiac/internal/terraform/cluster
Executing [ terraform destroy -var-file=/Users/przemek/git-repo/polinux/3rdparty/myiac/internal/terraform/cluster/cluster.tfvars -auto-approve ]
data.terraform_remote_state.state: Refreshing state...
google_compute_network.gke_network: Refreshing state... [id=projects/ozzy-playground/global/networks/ozzy-playground-dev-gke-network]
...
Destroy complete! Resources: 5 destroyed.


Kubernetes cluster deleted through Terraform

```

### CODE REVIEW

File: internal/cluster/cluster.go [hunk 1]
```diff
@@ -20,54 +27,69 @@ func InstallHelm() {
...
+	// Create Bucket if not present in the system
+	err := gcp.CreateGCSBucket(project, env)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// Check if terraform initialized
+	if _, err := os.Stat(tf+"/.terraform"); os.IsNotExist(err) {
+		fmt.Printf("Terraform not Initialized: in %v: Intializong now...", tf+"/.terraform\n")
+		argsArray := util.StringTemplateToArgsArray("%s", "init")
+		cmd := commandline.NewWithWorkingDir("terraform", argsArray, tf)
```
Comment (dfernandezm) [inline internal/cluster/cluster.go:40]: Use better `commandline.NewCommandLine(string)` which receives command as string

File: internal/cluster/cluster.go [hunk 2]
```diff
@@ -20,54 +27,69 @@ func InstallHelm() {
...
 
-//TODO: pass variables into the TF template/clustervars
-func CreateCluster(project string, env string, zone string) {
-	currentDir := util.CurrentExecutableDir()
-	tfFileLocation := currentDir + "/terraform/cluster"
-
-	varFileLoc := tfFileLocation + "/cluster.tfvars"
-	argsArray := util.StringTemplateToArgsArray("%s %s", "plan", "-var-file="+varFileLoc)
-	cmd := commandline.NewWithWorkingDir("terraform", argsArray, tfFileLocation)
+func PlanTerraform(tp string, tf string)  {
```
Comment (dfernandezm) [inline internal/cluster/cluster.go:45]: that's the idea, map `plan, apply, ...` will add a small doc with the Python module I have to translate. It has everything in it to manage terraform but that should go in `terraform.go`

File: internal/cluster/cluster.go (hunk 1 above)
Comment (dfernandezm) [reply internal/cluster/cluster.go:40]: That might not have `workingDir` actually, need another constructor :D

**Review by dfernandezm — CHANGES_REQUESTED** [review]
good start, minor changes

File: internal/gcp/bucket.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	for {
+		if bucketName == "" {
+			return fmt.Errorf("BucketName entered is empty %v.", bucketName)
+		}
+		attrs, err := buckets.Next()
+		// Assume bucket not found if at Iterator end and create
+		if err == iterator.Done {
+			// Create bucket
+			if err := bucket.Create(ctx, projectID, &storage.BucketAttrs{
+				Location: "EU",
```
[the remark this answers is not in the document]
Comment (pozgo) [reply internal/gcp/bucket.go:33]: Good point.

[the remark this answers is not in the document]
Comment (pozgo) [reply internal/gcp/bucket.go:35]: I'm Still thinking about the best approach to this. 

One would be to have a handler to read from `terraform` config file where the cluster is going to be deployed and have a function to return specific values that can be loaded there. Simple handler with all cases. 

I think for now I'm going to leave it to have it Hardcoded to EU. By default it's creating a multi-region bucket so no biggy if someone needs state files in another region. I'll put a `TODO:` there so we can update that later on. 
And will create an Issue too so we can pick this at some point and make GCS:Bucket fully customizable.

File: internal/cluster/cluster.go (hunk 1 above)
Comment (pozgo) [reply internal/cluster/cluster.go:40]: Yup. I'll keep it for now. There is more code-rewrite coming up soon. For now, I need operational repo.

File: internal/cluster/cluster.go (hunk 2 above)
Comment (pozgo) [reply internal/cluster/cluster.go:45]: Will implement it later. For now, I'll keep it. 

In the future, I'll consider using something like [this](https://github.com/getlunaform/go-terraform)

**Comment by pozgo** [comment]
I'm going to merge this. It's Working and now I can focus on a more direct approach per issue/feature at hand

File: internal/cluster/cluster.go (hunk 2 above)
Comment (dfernandezm) [reply internal/cluster/cluster.go]: Nice one you found it. Going to save so rewrites!!

**Pull request merged**